controllers/helmrelease_controller.go

```diff
@@ -128,271 +132,299 @@ func (c ConditionError) Error() string {
+		return result, err
 	}
-	return reconciledHr, ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, reconcileErr
+	return ctrl.Result{RequeueAfter: obj.Spec.Interval.Duration}, nil
 }
 
-type HelmReleaseReconcilerOptions struct {
-	MaxConcurrentReconciles   int
-	HTTPRetry                 int
-	DependencyRequeueInterval time.Duration
-}
-
-func (r *HelmReleaseReconciler) reconcileRelease(ctx context.Context,
-	hr v2.HelmRelease, chart *chart.Chart, values chartutil.Values) (v2.HelmRelease, error) {
+func (r *HelmReleaseReconciler) reconcileRelease(ctx context.Context, hr *v2.HelmRelease, chart *chart.Chart, values chartutil.Values) (ctrl.Result, error) {
 	log := logr.FromContext(ctx)
 
 	// Initialize Helm action runner
 	getter, err := r.getRESTClientGetter(ctx, hr)
 	if err != nil {
-		return v2.HelmReleaseNotReady(hr, v2.InitFailedReason, err.Error()), err
+		conditions.MarkFalse(hr, meta.ReadyCondition, v2.InitFailedReason, err.Error())
+		r.Eventf(ctx, hr, events.EventSeverityError, v2.InitFailedReason, err.Error())
+		return ctrl.Result{}, err
 	}
 	run, err := runner.NewRunner(getter, hr.GetStorageNamespace(), log)
 	if err != nil {
-		return v2.HelmReleaseNotReady(hr, v2.InitFailedReason, "failed to initialize Helm action runner"), err
+		conditions.MarkFalse(hr, meta.ReadyCondition, v2.InitFailedReason, "failed to initialize Helm action runner: %s", err)
+		r.Eventf(ctx, hr, events.EventSeverityError, v2.InitFailedReason, "failed to initialize Helm action runner: %s", err)
+		return ctrl.Result{}, err
 	}
 
 	// Determine last release revision.
-	rel, observeLastReleaseErr := run.ObserveLastRelease(hr)
+	rel, observeLastReleaseErr := run.ObserveLastRelease(*hr)
 	if observeLastReleaseErr != nil {
 		err = fmt.Errorf("failed to get last release revision: %w", observeLastReleaseErr)
-		return v2.HelmReleaseNotReady(hr, v2.GetLastReleaseFailedReason, "failed to get last release revision"), err
+		conditions.MarkFalse(hr, meta.ReadyCondition, v2.GetLastReleaseFailedReason, err.Error())
+		r.Eventf(ctx, hr, events.EventSeverityError, v2.GetLastReleaseFailedReason, err.Error())
+		return ctrl.Result{}, err
 	}
 
 	// Register the current release attempt.
 	revision := chart.Metadata.Version
 	releaseRevision := util.ReleaseRevision(rel)
 	valuesChecksum := util.ValuesChecksum(values)
-	hr, hasNewState := v2.HelmReleaseAttempted(hr, revision, releaseRevision, valuesChecksum)
-	if hasNewState {
-		hr = v2.HelmReleaseProgressing(hr)
-		if updateStatusErr := r.patchStatus(ctx, &hr); updateStatusErr != nil {
-			log.Error(updateStatusErr, "unable to update status after state update")
-			return hr, updateStatusErr
-		}
-		// Record progressing status
-		r.recordReadiness(ctx, hr)
-	}
+	// TODO: Progressing is set immediately on reconcile so is it necessary to set it here?
+	hr, _ = v2.HelmReleaseAttempted(hr, revision, releaseRevision, valuesChecksum)
```

Comment:
Should hasNewState be used here?

Reply:
To keep most of the logic (that should be rewritten as well as noted in (an) other issue(s)) intact I would maintain it for now.